### ISSUE TITLE
BABEL_TRANSFORM_ERROR - babel traverse context -  cannot read property "removeBinding" of null

### ISSUE THREAD

**Issue by salvoravida** [babel/babel#12383]
## Bug Report

**EDIT by @nicolo-ribaudo**

You can workaround this bug by making sure that `@babel/core` uses an older `@babel/traverse` version.
If you are using `yarn` you can specify `"@babel/traverse": "7.12.0"` in [`resolutions`](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/), otherwise you can try downgrading it as described by https://github.com/babel/babel/issues/12383#issuecomment-732049261.

---

https://github.com/babel/babel/pull/12331 after this pr

<details>

"This is technically a breaking change (we are changing the default value of a parameter in the public API), but I couldn't find anyone online using that parameter and I'm not sure why anyway would not wont it to use the correct context."

is there anyone that do code review, here @babel ?

What do you think about semver?

</details>

![image](https://user-images.githubusercontent.com/20126259/99918310-4ed45200-2d16-11eb-89b4-58597f3c84f7.png)

**Comment by nicolo-ribaudo** [comment]
The goal of #12331 was to fix problems _really similar_ to the one reported in your stack trace.

Plugins can optimize AST traversal by passing a `noScope: true` option, which marks the internal scope tracker as not needed and sets the internal scope tracker to `null`.

The problem (fixed by #12331) was that sometimes an old value of `noScope` was used, and if a plugin run with `noScope: true` and then a different plugin run with `noScope: false`, the second one didn't get a scope tracker instance causing errors such as `Cannot read property removeBinding of null`.

The "technically a breaking change" introduced by that PR is that now the `noScope` option is not ignored, so if a plugin sets `noScope: true` but then tries to use the scope tracker it will now throw an error (usually `noScope: true` was ignored before that PR). So the actual breaking change that I know of (and that I knew of when I wrote that description) is... respecting the value explicitly set by an option.

I'm thus surprised by this error, but yeah we should fix the regression.

However, we'll need your help to do it. When opening a new issue in this repository, GitHub gives you a form similar to this one:

![Schermata da 2020-11-23 02-21-57](https://user-images.githubusercontent.com/7000710/99922809-bdbfa400-2d32-11eb-981f-d4efd7b38d48.png)

As you can see, there are some clearly defined sections:
- **Current behavior**: :heavy_check_mark: (this is the stack trace you posted)
- **Input Code**: :x: 
- **Expected behavior**: :heavy_check_mark: (not crashing)
- **Babel Configuration**: :x:

Without the missing information (marked with :x:), we cannot fix this issue.
- If we just revert that PR, we don't have any guarantee that it actually fixes the error.
- If we are lucky and it fixes the error, we still cannot write a test for it. This makes it quite likely that we'll break it again when fixing in another way the bug that #12331 was originally resolving.

I understand that depending on your employer's policy it might be impossible for you to just give us the full contents of `CustomDatePicker.js`. You can comment out pieces of that file until the error goes away, and then just sharing the minimal portion of code needed to reproduce the crash.

**Comment by ehoogeveen** [comment]
I think `babel-preset-minify-simplify` is triggering this error (whether the fault is with the plugin or babel itself I don't know). It can be seen when minifying bowser v1.9.4 for example.

To reproduce:
1. Using `node.js LTS v14.15.1` with `npm version 6.14.8`, set up a clean node project.
1. Run `npm install --save-dev @babel/cli @babel/core babel-preset-minify bowser@1.9.4`
2. Run `npx babel --presets=babel-preset-minify node_modules/bowser/bowser.js`

**Comment by salvoravida** [comment]
@nicolo-ribaudo 
Are you serious? 
Do you understand that You have **_intentionally_** introduce a PR with a breaking change, without a major version, in a project used by 90% of the modern web?
I do not think so.

**Comment by stefaniapedrazzi** [comment]
We are also having issues due to this change and our CI is failing since then.

As far as I understand, but maybe it is wrong, there are conflicts with the babel minify presets that we solved by using the `--presets "@babel/preset-env"` command option.
Unfortunately this trick is no longer working and it throws the error mentioned in the issue description.

I understand that it could take a while to check and fix it.
But it would be very welcome if you could suggest a workaround.

**Comment by ehoogeveen** [comment]
A workaround that mostly worked for me was to disable the `minify-simplify` plugin by adding `["minify", { simplify: false }]` to the preset options. Unfortunately I then got a "Maximum call stack size exceeded" error when minifying d3 v4.13.0 (but I haven't investigated that yet).

**Comment by nicolo-ribaudo** [comment]
Thanks @ehoogeveen for the example and workaround! I will investigate a fix today, in the meantime you can add `@babel/traverse@7.12.0` to your `devDependencies`, uninstall `@babel/core` and reinstall it (so that it uses the older `@babel/traverse` version).

**Comment by nicolo-ribaudo** [comment]
I trimmed down the reproduction example to just two plugins: `babel-plugin-transform-merge-sibling-variables` and `babel-plugin-minify-simplify`. As an alternative to `["minify", { simplify: false }]`, a third workaround could be `["minify", { mergeVars: false }]`.

**Comment by magestic0100** [comment]
It works great if you keep original binary code. It is not to be modified. It is a demo default only. It is the backup reset configuration, backed by microsoft office, azure, IAM Data Management(google & IBM), AKA.MWCLOUD(AMAZON).

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by w3nda** [comment]
Having our CI build failing as well with the very same issue, without explicitly using any of the packages you mentioned here ...
is there any ETA for this issue? 

I can hardly tell which workaround could be used in my case, here are the babel packages we use

```
"@babel/cli": "^7.0.0"
"@babel/core": "^7.0.0"
"@babel/plugin-proposal-class-properties": "^7.0.0"
"@babel/plugin-proposal-object-rest-spread": "^7.0.0"
"@babel/plugin-syntax-dynamic-import": "^7.0.0"
"@babel/plugin-transform-modules-commonjs": "^7.0.0"
"@babel/plugin-transform-runtime": "^7.0.0"
"@babel/preset-env": "^7.0.0"
"@babel/preset-react": "^7.0.0"
```

And a snippet from our failing build - 
```
 Module build failed (from ./node_modules/babel-loader/lib/index.js):

ERR! TypeError: [OMITTED]: Cannot read property 'removeBinding' of null

ERR!     at Object.keys.forEach.name ([OMITTED]/node_modules/@babel/traverse/lib/path/removal.js:45:52)
```

**Comment by nicolo-ribaudo** [comment]
I'll open a PR today, hoping for a patch release today.

**Comment by nicolo-ribaudo** [comment]
I managed to reproduce the bug even reverting #12331: that PR made this bug more common, but didn't _cause_ it.

**Comment by nicolo-ribaudo** [comment]
Can the people who see this bug in their project try to replace the `getSibling` function in `node_modules/@babel/core/node_modules/@babel/traverse/lib/path/family.js` and/or `node_modules/@babel/traverse/lib/path/family.js` with this, and see if they still their build still crashes? I have identified the bug and fixed the example provided in https://github.com/babel/babel/issues/12383#issuecomment-732018914, but I want to make sure I'm not missing something.

```diff
  function getSibling(key) {
    return _index.default.get({
      parentPath: this.parentPath,
      parent: this.parent,
      container: this.container,
      listKey: this.listKey,
      key: key
-   });
+   }).setContext(this.context);
  }
```

**Comment by ehoogeveen** [comment]
With that change applied, the removeBinding errors are gone.

However I do still get the following error when minifying `sortablejs@1.10.2` (also new with babel 7.12.7):
```
RangeError: node_modules/sortablejs/Sortable.js: Maximum call stack size exceeded
    at NodePath._resolve (node_modules/@babel/traverse/lib/path/introspection.js:310:18)
    at NodePath.resolve (node_modules/@babel/traverse/lib/path/introspection.js:307:15)
    at NodePath._resolve (node_modules/@babel/traverse/lib/path/introspection.js:317:31)
    at NodePath.resolve (node_modules/@babel/traverse/lib/path/introspection.js:307:15)
    at node_modules/@babel/traverse/lib/path/inference/inferer-reference.js:72:27
    at Array.filter (<anonymous>)
    at getConstantViolationsBefore (node_modules/@babel/traverse/lib/path/inference/inferer-reference.js:71:21)
    at getTypeAnnotationBindingConstantViolations (node_modules/@babel/traverse/lib/path/inference/inferer-reference.js:36:28)
    at NodePath._default (node_modules/@babel/traverse/lib/path/inference/inferer-reference.js:22:14)
    at NodePath._getTypeAnnotation (node_modules/@babel/traverse/lib/path/inference/index.js:59:20) {
  code: 'BABEL_TRANSFORM_ERROR'
```
I also got that error without the fix when disabling minify-simplify. I'm not sure the stack is entirely consistent; it might be hitting the limit at different points (or staying just beneath it in some cases) - without the fix but with minify-simplify disabled, I got the error while minifying `d3@4.13.0`:
```
RangeError: output/js/ext/full/d3.js: Maximum call stack size exceeded
    at NodePath.setScope (node_modules/@babel/traverse/lib/path/context.js:121:18)
    at NodePath.setContext (node_modules/@babel/traverse/lib/path/context.js:149:8)
    at NodePath._getKey (node_modules/@babel/traverse/lib/path/family.js:212:8)
    at NodePath.get (node_modules/@babel/traverse/lib/path/family.js:186:17)
    at NodePath._resolve (node_modules/@babel/traverse/lib/path/introspection.js:316:14)
    at NodePath.resolve (node_modules/@babel/traverse/lib/path/introspection.js:307:15)
    at node_modules/@babel/traverse/lib/path/inference/inferer-reference.js:72:27
    at Array.filter (<anonymous>)
    at getConstantViolationsBefore (node_modules/@babel/traverse/lib/path/inference/inferer-reference.js:71:21)
    at getTypeAnnotationBindingConstantViolations (node_modules/@babel/traverse/lib/path/inference/inferer-reference.js:36:28) {
  code: 'BABEL_TRANSFORM_ERROR'
}
```
Tested on Windows 10 x64.

**Comment by stefaniapedrazzi** [comment]
> Can the people who see this bug in their project try to replace the getSibling function in node_modules/@babel/core/node_modules/@babel/traverse/lib/path/family.js and/or node_modules/@babel/traverse/lib/path/family.js with this, and see if they still their build still crashes?

I confirm that patching `getSibling` in `node_modules/@babel/core/node_modules/@babel/traverse/lib/path/family.js` fixes the issue in my project.

**Comment by nicolo-ribaudo** [comment]
@ehoogeveen Can you share that `d3.js` file? I tried with `npx babel --presets=babel-preset-minify node_modules/d3/build/d3.js` (which is [this file](https://unpkg.com/browse/d3@4.13.0/build/d3.js)) but I cannot reproduce the error.

**Comment by ehoogeveen** [comment]
I was able to work around the stack overflow by calling the babel cli using `node --stack-size=1536 node_modules/@babel/cli/bin/babel --presets=babel-preset-minify node_modules/sortablejs/Sortable.js` (for the sortablejs example), so it isn't infinite recursion.

**Comment by mhbarros** [comment]
> Thanks @ehoogeveen for the example and workaround! I will investigate a fix today, in the meantime you can add `@babel/traverse@7.12.0` to your `devDependencies`, uninstall `@babel/core` and reinstall it (so that it uses the older `@babel/traverse` version).

Thanks, it worked just fine. I also updated my package-lock.json with these values and changed my CI system to use the command "npm ci" instead of "npm install", to always pull this version until i manually update (thanks @nicolo-ribaudo for the tip!)

**Comment by nicolo-ribaudo** [comment]
I'm keeping this open until it's released, in case someone else has the same problem.

**Comment by brunomptorres** [comment]
> I'm keeping this open until it's released, in case someone else has the same problem.

@nicolo-ribaudo any idea on when this is getting released?

**Comment by nicolo-ribaudo** [comment]
Right after https://github.com/babel/babel/pull/12390 is merged, which fixes another regression with this release.

**Comment by nicolo-ribaudo** [comment]
Released as `7.12.8`
